File: 0001~0500/0095. Unique Binary Search Trees II/main.cpp
/** 
 * Source: https://leetcode.com/problems/unique-binary-search-trees-ii/
 * Date: 2021/7/16
 * Skill: Recursion
 * Recursion: Trees(n) = Trees(0)*Trees(n-1) + Trees(1)*Trees(n-2) + … + Tress(n-1)*Trees(0)
 * Runtime: 16 ms, faster than 79.00% of C++ online submissions
 * Memory Usage: 16.7 MB, less than 5.80% of C++ online submissions
 * Time complexity: O(3^n)
 * Space complexity: O(3^n)
 * Constraints: 
 *      1 <= n <= 8
 */

#include<iostream>
#include<vector>
#include<string>
using namespace std;


// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    vector<TreeNode*> generateTrees(int n) {
        return generateTrees(1,n);
    }

    vector<TreeNode*> generateTrees(int l, int r) {
        vector<TreeNode*> ans;
        if (l>r) 
            return {NULL};
        for (int index = l; index <= r; index++) {
            for(auto left: generateTrees(l, index-1)) {
                for(auto right: generateTrees(index+1, r)) {
                    ans.push_back(new TreeNode(index));
                    ans.back()->left = left;
                    ans.back()->right = right;
                }
            }
        }
        return ans;
    }


};

static const auto io_sync_off = []() {
    // turn off sync
    std::ios::sync_with_stdio(false);
    // untie in/out streams
    std::cin.tie(nullptr);
    return nullptr;
}();

File: 0001~0500/0095. Unique Binary Search Trees II/main.py
"""
Source: https://rb.gy/ypwoj
Date: 2023/8/5
Skill:
Ref:
Runtime: 65 ms, faster than 74.74%
Memory Usage: 18.53 MB, less than 7.09%
Time complexity:
Space complexity:
Constraints:
    1 <= n <= 8
"""

import math
from typing import List, Tuple, Optional
from collections import defaultdict, deque
from heapq import heapify, heappush, heappop, nsmallest, heappushpop
import bisect
import heapq
import functools


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
        def gen_tree(nums):
            if len(nums) == 0:
                return [None]
            sz, res = len(nums), []
            for i in range(sz):
                for left_tree in gen_tree(nums[:i]):
                    for right_tree in gen_tree(nums[i + 1:]):
                        new_root = TreeNode(val=nums[i], left=left_tree, right=right_tree)
                        res.append(new_root)
            return res

        return gen_tree([i + 1 for i in range(n)])


if __name__ == "__main__":
    s = Solution()
